packages/byods/src/BYODS.ts

```diff
@@ -1,47 +1,389 @@
+      throw error;
+    }
   }
 
-  public async makeAuthenticatedRequest(endpoint: string): Promise<any> {
-    if (new Date() >= new Date(this.expiresAt as Date)) {
-      throw new Error('Token has expired');
+  /**
+   * Creates a service app authorization.
+   *
+   * @param orgId - The organization ID.
+   * @param personalAccessToken - The personal access token of the developer of the service app.
+   * @returns A promise that resolves to void.
+   * @throws If the request fails or an error occurs during the process.
+   */
+  public async createServiceAppAuthorization(
+    orgId: string,
+    personalAccessToken: string
+  ): Promise<TokenResponse> {
+    try {
+      const response = await fetch(`${this.baseUrl}/applications/${this.serviceAppId}/token`, {
+        method: 'POST',
+        headers: {
+          Authorization: `Bearer ${personalAccessToken}`,
+          'Content-Type': 'application/json',
+          ...this.headers,
+        },
+        body: new URLSearchParams({
+          targetOrgId: orgId,
+          clientId: this.config.clientId,
+          clientSecret: this.config.clientSecret,
+        }),
+      });
+
+      if (!response.ok) {
+        throw new Error(`Failed to create service app authorization: ${response.statusText}`);
+      }
+
+      const data: TokenResponse = (await response.json()) as TokenResponse;
+      this.updateServiceAppToken(data, orgId);
+
+      return data;
+    } catch (error) {
+      console.error('Error creating service app authorization:', error);
+      throw error;
     }
+  }
+
+  /**
+   * Retrieve a new service app token after authorization.
+   * @param {string} orgId - The organization ID.
+   * @param {string} token - The service app owner's personal access token or token from an integration that has the scope `spark:applications_token`.
+   * @returns {Promise<void>}
+   */
+  public async retrieveTokenAfterAuthorization(orgId: string, token: string): Promise<void> {
+    try {
+      const response = await fetch(`${this.baseUrl}/applications/${this.serviceAppId}/token`, {
+        method: 'POST',
+        headers: {
+          Authorization: `Bearer ${token}`,
+          'Content-Type': 'application/x-www-form-urlencoded',
+          ...this.headers,
+        },
+        body: new URLSearchParams({
+          targetOrgId: orgId,
+          clientId: this.config.clientId,
+          clientSecret: this.config.clientSecret,
+        }),
+      });
 
-    // Use this.token.access_token to make authenticated requests
-    const response = await fetch(endpoint, {
+      if (!response.ok) {
+        throw new Error(`Failed to retrieve token: ${response.statusText}`);
+      }
+
+      const data: TokenResponse = (await response.json()) as TokenResponse;
+      this.updateServiceAppToken(data, orgId);
+    } catch (error) {
+      console.error('Error retrieving token after authorization:', error);
+      throw error;
+    }
+  }
```

Comment:
The above 3 methods are doing the same thing. Do you think it would be better to create a single method that takes in the url and body and makes that request?

Reply:
refreshServiceAppAccessToken, retrieveTokenAfterAuthorization, createServiceAppAuthorization, saveServiceAppRegistration all 4 have the same structure.